### PR TITLE
[SPARK-19021][YARN] Generailize HDFSCredentialProvider to support non HDFS security filesystems

### PR DESCRIPTION
**Pull request by jerryshao**
Currently Spark can only get token renewal interval from security HDFS (hdfs://), if Spark runs with other security file systems like webHDFS (webhdfs://), wasb (wasb://), ADLS, it will ignore these tokens and not get token renewal intervals from these tokens. These will make Spark unable to work with these security clusters. So instead of only checking HDFS token, we should generalize to support different DelegationTokenIdentifier.

## How was this patch tested?

Manually verified in security cluster.

### CODE REVIEW

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by tgravescs** [comment]
what kind of testing have you done on this? 

A couple of concerns:  

1) FileSystem is kind of obscure, really I think it means hadoop filesystems.  It won't work for just any filesystem.

2) this is documented: https://spark.apache.org/docs/latest/running-on-yarn.html  I'm not fond of changing config names.  Its basically an api.  However, I doubt many people are specifically setting it at this point though since its in the defaults.  Either way we would need to update that documentation.

**Comment by jerryshao** [comment]
Thanks @tgravescs for your comments. Yes actually it should be Hadoop compatible file systems, or in other words should be Hadoop `FileSystem` API supported FS. I tested with kerberized HDFS and wasb.

In HDI, it ships with kerberized wasb or ADSL, they will also have delegation token but with different token kind. In the current code we assume that token kind should be `HDFS_DELEGATION_KIND`, this is not true if the default FS is not HDFS or there's no HDFS in the cluster. In such case currently we simply ignore the tokens from these FS, which will be an issue for long running Spark application. This also applies to webHDFS (it also has different token kind).

**Comment by tgravescs** [comment]
Ok can we change the name of this from FileSystem to be HadoopFileSystem or HadoopFs to be a bit more clear and then update the docs accordingly to make sure we specify what the requirements are for that to work on those filesystems.

File: docs/running-on-yarn.md [hunk 1]
```diff
@@ -479,12 +479,12 @@ Hadoop services issue *hadoop tokens* to grant access to the services and data.
 Clients must first acquire tokens for the services they will access and pass them along with their
 application as it is launched in the YARN cluster.
 
-For a Spark application to interact with HDFS, HBase and Hive, it must acquire the relevant tokens
+For a Spark application to interact with Hadoop filesystem, HBase and Hive, it must acquire the relevant tokens
```
Comment (tgravescs) [inline docs/running-on-yarn.md]: I think perhaps we should say interact with any of the Hadoop filesystems (for example hdfs, webhdfs, etc).

File: docs/running-on-yarn.md [hunk 2]
```diff
@@ -479,12 +479,12 @@ Hadoop services issue *hadoop tokens* to grant access to the services and data.
...
 using the Kerberos credentials of the user launching the application
 —that is, the principal whose identity will become that of the launched Spark application.
 
 This is normally done at launch time: in a secure cluster Spark will automatically obtain a
-token for the cluster's HDFS filesystem, and potentially for HBase and Hive.
+token for the cluster's Hadoop filesystem, and potentially for HBase and Hive.
```
Comment (tgravescs) [inline docs/running-on-yarn.md]: Perhaps say the default hadoop filesystem

File: docs/running-on-yarn.md [hunk 3]
```diff
@@ -494,7 +494,7 @@ Similarly, a Hive token will be obtained if Hive is on the classpath, its config
 includes a URI of the metadata store in `"hive.metastore.uris`, and
 `spark.yarn.security.credentials.hive.enabled` is not set to `false`.
 
-If an application needs to interact with other secure HDFS clusters, then
+If an application needs to interact with other secure Hadoop filesystem clusters, then
```
Comment (tgravescs) [inline docs/running-on-yarn.md]: remove "clusters" here.  should just say "interact with other secure Hadoop filesystems".

Comment (tgravescs) [inline docs/running-on-yarn.md]: can you also change the example below to include webhdfs

**Comment by tgravescs** [comment]
few minor doc comments, otherwise looks good. 

Could you also please test with webhdfs.

**Comment by jerryshao** [comment]
@tgravescs thanks for your comments, I also tested with webhdfs with configuration like:

```
--conf spark.yarn.access.namenodes=webhdfs://c6402.ambari.apache.org:50070/
```

And the related logs are:

```
17/01/09 03:15:49 INFO HadoopFSCredentialProvider: getting token for: webhdfs://c6402.ambari.apache.org:50070/
17/01/09 03:15:50 INFO HadoopFSCredentialProvider: getting token for: hdfs://c6402.ambari.apache.org:8020/user/spark
17/01/09 03:15:50 INFO DFSClient: Created HDFS_DELEGATION_TOKEN token 138 for spark on 192.168.64.102:8020
17/01/09 03:15:50 INFO DFSClient: Created HDFS_DELEGATION_TOKEN token 140 for spark on 192.168.64.102:8020
17/01/09 03:15:50 INFO HadoopFSCredentialProvider: Renewal interval is 86400112 for token WEBHDFS delegation
17/01/09 03:15:50 INFO HadoopFSCredentialProvider: Renewal interval is 86400146 for token HDFS_DELEGATION_TOKEN
```

As you could see also get tokens and token renewal interval from webhdfs token.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala [hunk 4]
```diff
@@ -56,14 +57,14 @@ private[security] class HDFSCredentialProvider extends ServiceCredentialProvider
...
+      val nextRenewalDates = creds.getAllTokens.asScala
+        .filter(_.decodeIdentifier().isInstanceOf[AbstractDelegationTokenIdentifier])
         .map { t =>
-          val identifier = new DelegationTokenIdentifier()
-          identifier.readFields(new DataInputStream(new ByteArrayInputStream(t.getIdentifier)))
+          val identifier = t.decodeIdentifier().asInstanceOf[AbstractDelegationTokenIdentifier]
           identifier.getIssueDate + interval
-      }.foldLeft(0L)(math.max)
+        }
+      if (nextRenewalDates.isEmpty) None else Some(nextRenewalDates.min)
```
Comment (viirya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala]: Why previously we use math.max and now changed to minimum value?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala [hunk 5]
```diff
@@ -56,14 +57,14 @@ private[security] class HDFSCredentialProvider extends ServiceCredentialProvider
     }
 
     // Get the time of next renewal.
-    tokenRenewalInterval.map { interval =>
-      creds.getAllTokens.asScala
-        .filter(_.getKind == DelegationTokenIdentifier.HDFS_DELEGATION_KIND)
+    tokenRenewalInterval.flatMap { interval =>
+      val nextRenewalDates = creds.getAllTokens.asScala
+        .filter(_.decodeIdentifier().isInstanceOf[AbstractDelegationTokenIdentifier])
```
Comment (viirya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala]: Can we make sure all instances of `AbstractDelegationTokenIdentifier` are tokens for Hadoop file systems? Isn't Hive's `DelegationTokenIdentifier` also an `AbstractDelegationTokenIdentifier`?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala (hunk 4 above)
Comment (jerryshao) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala]: I think the logic was incorrect previously, we should get the minimum renewal date, that's why I changed here. But the code was probably OK if only we only have one default HDFS configured.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala (hunk 5 above)
Comment (jerryshao) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala]: Yeah, that's the issue, let me think about how to handle it.

**Comment by jerryshao** [comment]
@tgravescs , would you please review again, thanks a lot.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala [hunk 6]
```diff
@@ -17,37 +17,39 @@
...
       sparkConf: SparkConf,
       creds: Credentials): Option[Long] = {
     // NameNode to access, used to get tokens from different FileSystems
+    val tmpCreds = new Credentials()
     nnsToAccess(hadoopConf, sparkConf).foreach { dst =>
       val dstFs = dst.getFileSystem(hadoopConf)
-      logInfo("getting token for namenode: " + dst)
-      dstFs.addDelegationTokens(getTokenRenewer(hadoopConf), creds)
+      logInfo("getting token for: " + dst)
+      dstFs.addDelegationTokens(getTokenRenewer(hadoopConf), tmpCreds)
```
Comment (viirya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala]: Can you put `getTokenRenewer` outside the foreach loop and use a variable for it? We don't need to get the token renewer everytime.

Comment (jerryshao) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HadoopFSCredentialProvider.scala]: Done, thanks for your comment.

**Comment by tgravescs** [comment]
so the changes look good.  I just thought of one other thing though and that is really the name and description of spark.yarn.access.namenodes is off.   I'll file a separate jira for that though since unrelated.

**Pull request closed without merging**